### PR TITLE
support resource level rule replacement in circuitBreaker module

### PR DESCRIPTION
**Pull request by sanxun0325**
### Describe what this PR does / why we need it
Support resource level rule replacement in circuitBreaker module.

### Special notes for reviews

### CODE REVIEW

File: core/circuitbreaker/rule_manager.go [hunk 1]
```diff
@@ -132,16 +132,52 @@ func ClearRules() error {
...
 		logging.Info("[CircuitBreaker] Load rules is the same with current rules, so ignore load operation.")
 		return false, nil
 	}
 
-	err := onRuleUpdate(rules)
+	err := onRuleUpdate(resRulesMap)
+	return true, err
+}
+
+func LoadRulesOfResource(res string, rules []*Rule) (bool, error) {
```
Comment (luckyxiaoqiang) [inline core/circuitbreaker/rule_manager.go:156]: Better to add comments for public API.

File: core/circuitbreaker/rule_manager.go [hunk 2]
```diff
@@ -204,84 +230,38 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
-
-			if reuseStatIdx >= 0 {
-				breakersClone[res] = append(oldResCbs[:reuseStatIdx], oldResCbs[reuseStatIdx+1:]...)
-			}
-			insertCbToCbMap(cb, res, newBreakers)
+	newBreakers := make(map[string][]CircuitBreaker, len(validResRulesMap))
+	for res, resRules := range validResRulesMap {
+		newCbsOfRes := buildResourceCircuitBreaker(res, resRules, breakersClone[res])
+		if len(newCbsOfRes) > 0 {
+			newBreakers[res] = newCbsOfRes
```
Comment (luckyxiaoqiang) [inline core/circuitbreaker/rule_manager.go:264]: Maybe we can merge with the below loop.

File: core/circuitbreaker/rule_manager.go (hunk 1 above)
Comment (sanxun0325) [reply core/circuitbreaker/rule_manager.go:158]: done

File: core/circuitbreaker/rule_manager.go (hunk 2 above)
Comment (sanxun0325) [reply core/circuitbreaker/rule_manager.go:266]: done

File: core/circuitbreaker/rule_manager.go [hunk 3]
```diff
@@ -204,103 +232,102 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
 
-			generator := cbGenFuncMap[r.Strategy]
-			if generator == nil {
-				logging.Warn("[CircuitBreaker onRuleUpdate] Ignoring the rule due to unsupported circuit breaking strategy", "rule", r)
-				continue
-			}
+	updateMux.Lock()
+	breakerRules = validResRulesMap
+	breakers = newBreakers
+	currentRules = rawResRulesMap
```
Comment (liqiangz) [inline core/circuitbreaker/rule_manager.go:276]: We can do this without lock.

Comment (sanxun0325) [reply core/circuitbreaker/rule_manager.go]: done

**Review by luckyxiaoqiang — APPROVED** [review]
LGTM

**Review by liqiangz — APPROVED** [review]
LGTM

**Review by louyuting — APPROVED** [review]
LGTM

**Pull request merged**